ethexe/contracts/src/Router.sol

```diff
@@ -485,25 +485,45 @@ contract Router is IRouter, OwnableUpgradeable, ReentrancyGuardTransientUpgradea
         return Gear.validatorsCommitmentHash(_commitment);
     }
 
+    function _checkTransitionsSortedByValueSign(Gear.StateTransition[] calldata _transitions) private pure {
+        bool isPositiveFound = false;
+
+        for (uint256 i = 0; i < _transitions.length; i++) {
+            // If negative sign found after positive sign - transitions are not sorted.
+            if (!_transitions[i].valueToReceiveNegativeSign) {
+                isPositiveFound = true;
+                continue;
+            }
+
+            require(
+                !(isPositiveFound && _transitions[i].valueToReceiveNegativeSign),
+                "state transitions must be sorted by valueToReceiveNegativeSign in descending order"
+            );
+        }
+    }
+
     function _commitTransitions(Storage storage router, Gear.StateTransition[] calldata _transitions)
         private
         returns (bytes32)
     {
         bytes memory transitionsHashes;
 
+        _checkTransitionsSortedByValueSign(_transitions);
```

Comment:
You don't need to iterate it all. I may have here mut bool and on each transition check that sign changed just once